js/messaging.js

```diff
@@ -232,65 +236,51 @@ function findExistTab({url, currentWindow, ignoreHash = true, ignoreSearch = fal
+ * @param {number} [_.openerTabId] defaults to the active tab
+ * @param {Boolean} [_.active=true] `true` to activate the tab
+ * @param {Boolean|null} [_.currentWindow=true] `null` to check all windows
+ * @param {Boolean} [_.newWindow=false] `true` to open a new window
+ * @param {chrome.windows.CreateData} [_.windowPosition] options for chrome.windows.create
+ * @returns {Promise<chrome.tabs.Tab>} Promise -> opened/activated tab
  */
-function openURL(options) {
-  if (typeof options === 'string') {
-    options = {url: options};
-  }
-  let {
-    url,
-    index,
-    active,
-    currentWindow = true,
-    newWindow = false,
-    windowPosition
-  } = options;
-
+function openURL({
+  url,
+  index,
+  openerTabId,
+  active = true,
+  currentWindow = true,
+  newWindow = false,
+  windowPosition,
+}) {
   if (!url.includes('://')) {
     url = chrome.runtime.getURL(url);
   }
-  return findExistTab({url, currentWindow}).then(tab => {
+  return findExistingTab({url, currentWindow}).then(tab => {
     if (tab) {
-      // update url if only hash is different?
-      // we can't update URL if !url.includes('#') since it refreshes the page
-      // FIXME: should we move the tab (i.e. specifying index)?
-      if (tab.url !== url && tab.url.split('#')[0] === url.split('#')[0] &&
-          url.includes('#')) {
-        return activateTab(tab, {url, index});
-      }
-      return activateTab(tab, {index});
-    }
-    if (newWindow) {
-      return createWindow(Object.assign({url}, windowPosition));
+      return activateTab(tab, {
+        index,
+        openerTabId,
+        // when hash is different we can only set `url` if it has # otherwise the tab would reload
+        url: url !== tab.url && url.includes('#') && url,
+      });
+    } else if (newWindow && createWindow) {
+      return createWindow(Object.assign({url}, windowPosition))
+        .then(wnd => wnd.tabs[0]);
     }
-    return getActiveTab().then(tab => {
-      if (isTabReplaceable(tab, url)) {
-        // don't move the tab in this case
-        return activateTab(tab, {url});
-      }
-      const options = {url, index, active};
-      // FF57+ supports openerTabId, but not in Android (indicated by the absence of chrome.windows)
-      // FIXME: is it safe to assume that the current tab is the opener?
-      if (tab && !tab.incognito && (!FIREFOX || FIREFOX >= 57 && chrome.windows)) {
-        options.openerTabId = tab.id;
-      }
-      return createTab(options);
-    });
+    return getActiveTab().then((activeTab = {url: ''}) =>
+      isTabReplaceable(activeTab, url) ?
+        activateTab(activeTab, {url, openerTabId}) : // not moving the tab
+        createTabWithOpener(activeTab, {url, index, active}));
   });
+  function createTabWithOpener(openerTab, options) {
+    if (openerTabIdSupported && !openerTab.incognito) {
+      options.openerTabId = openerTabId == null ? openerTab.id : openerTabId;
```

Comment:
`openerTab.id` can be `undefined`. Is it fine to pass `undefined` as `openerTabId` to `createTab`?

Reply:
It's always been fine but they can break it any time, right? So I'll add a check.